### PR TITLE
Add a remove method to ListenerCollection

### PR DESCRIPTION
**Pull request by esanzgar**
This new method allows to unregister individual event listeners.

### CODE REVIEW

File: src/shared/listener-collection.js
Comment (robertknight) [inline src/shared/listener-collection.js:19]: There should be a blank line between the description of a symbol and any JSDoc tags (`@param`, `@prop` etc.).

File: src/shared/test/listener-collection-test.js
Comment (robertknight) [inline src/shared/test/listener-collection-test.js:22]: Can you wrap the tests here in `describe(<method name>)` blocks to make it more obvious which functions are being tested, eg:

```js
describe('#remove', () => {
  it('removes an event listener', () => { ... });
});

describe('#removeAll', () => {
  it('unregisters all event listeners', () => { ... });
});
```

File: src/shared/listener-collection.js
Comment (robertknight) [inline src/shared/listener-collection.js:27]: The first listener ID is returned is 0, which is a falsey value. This could cause problems if calling code tries to do `if (!listenerId) { ... }`. I think it would be safer to make IDs start from 1, like `setTimeout` IDs.

File: src/shared/listener-collection.js
Comment (robertknight) [inline src/shared/listener-collection.js:39]: I see steps have been taken to avoid ID reuse here, which is good. It does mean that you have a growing array of empty slots to loop over in future though. Also `removeAll` will result in ID reuse.

File: src/shared/listener-collection.js
Comment (robertknight) [inline src/shared/listener-collection.js:50]: If `removeAll` is called and then new listeners are added, listener IDs will be reused.

File: src/shared/listener-collection.js
Comment (robertknight) [inline src/shared/listener-collection.js:14]: There is currently a listener ID reuse hazard in `removeAll` (see comments for that method). Although you could fix that while continuing to use an array, I think it would be more natural to use a map of listener ID to listener, together with an ID counter that doesn't get reset.

**Review by robertknight — CHANGES_REQUESTED** [review]
The proposed API looks good. However I spotted a couple of issues with the way listener IDs are allocated and stored internally. I also had some minor organizational feedback about the tests.

**Review by esanzgar — COMMENTED** [review]
As you suggested, I have reimplemented the `ListenerCollection` using internally a map. The keys (`listenerId`s) are symbols. Let me know if you find any pitfall in this alternative implementation.

File: src/shared/listener-collection.js
Comment (robertknight) [inline src/shared/listener-collection.js:30]: Using a symbol works. I guess you could also have made `this._listeners` a set which stores the `{ eventTarget, eventType, listener }` objects and return the object reference as a key.

File: src/shared/listener-collection.js
Comment (robertknight) [inline src/shared/listener-collection.js:35]:
```suggestion
   * Remove a listener using a listenerId
   *
   * @param {Symbol} listenerId
```

File: src/shared/test/listener-collection-test.js
Comment (robertknight) [inline src/shared/test/listener-collection-test.js:25]:
```suggestion
    it('unregisters the specified listener', () => {
```

File: src/shared/listener-collection.js
Comment (robertknight) [inline src/shared/listener-collection.js:19]:
```suggestion
   * Add a listener and return an ID that can be used to remove it later
```

**Review by robertknight — APPROVED** [review]
LGTM, once the earlier branch is merged.

**Pull request merged**
